File: Python/reverse_words_in_a_string_3.py
# https://leetcode.com/problems/reverse-words-in-a-string-iii/description/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def reverseWords(self, s: str) -> str:
        s = list(s)
        start: int = 0
        end: int = 0
        while end <= len(s):
            if end < len(s) and not ' ' == s[end]:
                end += 1
                continue
            end -= 1
            p: int = end
            while start < end:
                s[start], s[end] = s[end], s[start]
                start += 1
                end -= 1
            end = p + 2
            start = end
        return ''.join(s)

def examine()-> None:
    Examiner(Solution(), 'reverseWords').run((
        Testcase(('',), ''),
        Testcase(('ab',), 'ba'),
        Testcase(('Mr Ding',), 'rM gniD'),
        Testcase(("Let's take LeetCode contest",), "s'teL ekat edoCteeL tsetnoc"),
    ))

File: TypeScript/solutions/reverse-words-in-a-string-3.ts
// https://leetcode.com/problems/reverse-words-in-a-string-iii/description/
import { Examiner, Testcase } from '../utils';


function reverseWords(s: string): string {
    const arr = s.split('');
    let start = 0, end = 0;
    while (end <= arr.length) {
        if (end < arr.length && arr[end] !== ' ') { end++; continue; }
        let l = start, r = end - 1;
        while (l < r) {
            [arr[l], arr[r]] = [arr[r]!, arr[l]!];
            l++; r--;
        }
        end++;
        start = end;
    }
    return arr.join('');
}

export function examine(): void {
    new Examiner(reverseWords).run([
        new Testcase([''], ''),
        new Testcase(['ab'], 'ba'),
        new Testcase(['Mr Ding'], 'rM gniD'),
        new Testcase(["Let's take LeetCode contest"], "s'teL ekat edoCteeL tsetnoc"),
    ]);
}
